synapse/lib/drive.py

```diff
@@ -0,0 +1,536 @@
+import regex
+import asyncio
+
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.base as s_base
+import synapse.lib.config as s_config
+import synapse.lib.msgpack as s_msgpack
+import synapse.lib.schemas as s_schemas
+
+nameregex = regex.compile(s_schemas.re_drivename)
+
+LKEY_TYPE = b'\x00' # <type> = <schema>
+LKEY_DIRN = b'\x01' # <bidn> <name> = <kid>
+LKEY_INFO = b'\x02' # <bidn> = <info>
+LKEY_DATA = b'\x03' # <bidn> <vers> = <data>
+LKEY_VERS = b'\x04' # <bidn> <vers> = <versinfo>
+LKEY_INFO_BYTYPE = b'\x05' # <type> 00 <bidn> = 01
+
+rootdir = '00000000000000000000000000000000'
+
+def getVersIndx(vers):
+    maji = vers[0].to_bytes(3, 'big')
+    mini = vers[1].to_bytes(3, 'big')
+    pati = vers[2].to_bytes(3, 'big')
+    return maji + mini + pati
+
+class Drive(s_base.Base):
+    '''
+    Drive is a hierarchical storage abstraction which:
+
+    * Provides enveloping which includes meta data for each item:
+      * creator iden / time
+      * updated iden / time / version
+      * number of children
+      * data type for the item
+      * easy perms (enforcement is up to the caller)
+
+    * Enforces schemas for data
+    * Allows storage of historical versions of data
+    * Provides a "path traversal" based API
+    * Provides an iden based API that does not require traversal
+    '''
+    async def __anit__(self, slab, name):
+        await s_base.Base.__anit__(self)
+        self.slab = slab
+        self.dbname = slab.initdb(f'drive:{name}')
+        self.validators = {}
+
+    def getPathNorm(self, path):
+
+        if isinstance(path, str):
+            path = path.strip().strip('/').split('/')
+
+        return [p.strip().lower() for p in path]
+
+    def getItemInfo(self, iden):
+        return self._getItemInfo(s_common.uhex(iden))
+
+    def _getItemInfo(self, bidn):
+        byts = self.slab.get(LKEY_INFO + bidn, db=self.dbname)
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    def _reqItemInfo(self, bidn):
+        info = self._getItemInfo(bidn)
+        if info is not None:
+            return info
+
+        mesg = 'No drive item with ID {s_common.ehex(bidn)}.'
+        raise s_exc.NoSuchIden(mesg=mesg)
+
+    async def setItemPath(self, iden, path):
+        '''
+        Move an existing item to the given path.
+        '''
+        return await self._setItemPath(s_common.uhex(iden), path)
+
+    async def getItemPath(self, iden):
+        pathinfo = []
+        while iden is not None:
+
+            info = self.getItemInfo(iden)
+            if info is None:
+                break
+
+            pathinfo.append(info)
+            iden = info.get('parent')
+
+        pathinfo.reverse()
+        return pathinfo
+
+    async def _setItemPath(self, bidn, path):
+
+        path = self.getPathNorm(path)
+        pathinfo = await self.getPathInfo(path[:-1])
+
+        self.reqFreeStep(pathinfo[-1].get('iden'), path[-1])
```

Comment:
Should this be able to move items into the root dir? Currently doing something like `cell.setDriveInfoPath(iden, 'foo')` will raise an IndexError here

Reply:
great catch, thanks!